app/services/middleware_topology_service.rb

```diff
@@ -33,14 +36,28 @@ def entity_display_type(entity)
 
   def build_entity_data(entity)
     data = build_base_entity_data(entity)
-    data.merge!(:status => 'Unknown',
-                :display_kind => entity_display_type(entity))
-    data[:icon] = entity.decorate.try(:listicon_image)
+    data[:status] = 'Unknown'
+    data[:display_kind] = entity_display_type(entity)
+
+    # fonticon takes precedence before the listicon_image
+    # and both methods on decorator takes precedence before the global icons hash from the UiServiceMixin
+    data[:icon] = if !(icon = entity.decorate.try(:fonticon)).nil?
```

Comment:
I think, it's obvious from the comment. If the entity has a decorator and the decorator's `fonticon` method doesn't return `nil`, it's used, then the fallback is `listicon_image` and then it falls back to the entity type based approach using the `icons` (https://git.io/vrapO)